src/node_crypto.cc

```diff
@@ -1974,7 +1974,78 @@ static Local<Object> X509ToObject(Environment* env, X509* cert) {
 }
 
 
-// TODO(indutny): Split it into multiple smaller functions
+// Put issuer inside the object
+static void AddIssuer(X509** cert,
+                      const STACK_OF(X509)* const peer_certs,
+                      Local<Object>* const info,
+                      Environment* const env) {
+  *cert = sk_X509_delete(peer_certs, 0);
+  while (sk_X509_num(peer_certs) > 0) {
+    int i;
+    for (i = 0; i < sk_X509_num(peer_certs); i++) {
+      X509* ca = sk_X509_value(peer_certs, i);
+      if (X509_check_issued(ca, *cert) != X509_V_OK)
+        continue;
+
+      Local<Object> ca_info = X509ToObject(env, ca);
+      (*info)->Set(env->issuercert_string(), ca_info);
+      *info = ca_info;
+
+      // NOTE: Intentionally freeing cert that is not used anymore
+      X509_free(*cert);
+
+      // Delete cert and continue aggregating issuers
+      *cert = sk_X509_delete(peer_certs, i);
+      break;
+    }
+
+    // Issuer not found, break out of the loop
+    if (i == sk_X509_num(peer_certs))
+      break;
+  }
+}
+
+
+static bool CloneSSLCerts(X509** cert,
+                          STACK_OF(X509)** peer_certs,
+                          const STACK_OF(X509)* const ssl_certs) {
+  *peer_certs = sk_X509_new(nullptr);
+  if (*cert != nullptr)
+    sk_X509_push(*peer_certs, *cert);
+  for (int i = 0; i < sk_X509_num(ssl_certs); i++) {
+    *cert = X509_dup(sk_X509_value(ssl_certs, i));
+    if (*cert == nullptr)
+      return false;
+    if (!sk_X509_push(*peer_certs, *cert))
+      return false;
+  }
+  return true;
+}
```

Comment:
It might be nicer (certainly more Obviously Correct) to move the `sk_X509_pop_free()` on error into this function.

Reply:
I'll take a look and fix this. Thanks

Reply:
I've pushed a few commit but I've not addressed this yet. Will revisit this next week. I was wondering if you meant that I add sk_X509_pop_free() to this function in addition to the one in GetPeerCertificate which I think is needed as peer_cert is used after this function?